solidity/test/Oracle.js

```diff
@@ -0,0 +1,103 @@
+const { expect } = require("chai");
+const { expectRevert, expectEvent, BN, ether, constants } = require("@openzeppelin/test-helpers");
+const { latest, latestBlock } = require("@openzeppelin/test-helpers/src/time");
+
+const { ZERO_ADDRESS } = constants;
+
+const Oracle = artifacts.require("Oracle");
+
+contract("Oracle", (accounts) => {
+  let oracle;
+  let price0 = ether("1");
+  let price1 = ether("0.5");
+
+  const pool = accounts[1];
+
+  beforeEach(async () => {
+    oracle = await Oracle.new();
+    await oracle.setLiquidityPool(pool);
+  });
+
+  it("should revert if k value not set by owner", async () => {
+    await expectRevert(oracle.setK(10, { from: accounts[2] }), "ERR_ACCESS_DENIED");
+  });
+
+  it("should revert if k value do not lie within a valid range", async () => {
+    await expectRevert(oracle.setK(100000), "ERR_INVALID_K_VALUE");
+  });
+
+  it("should set k value", async () => {
+    const k = new BN("30");
+    const receipt = await oracle.setK(k);
+    expect(await oracle.k.call()).to.be.bignumber.equal(k);
+    expectEvent(receipt, "KValueUpdate", {
+      _k: k
+    });
+  });
+
+  it("should revert if pool address is not set by owner", async () => {
+    await expectRevert(oracle.setLiquidityPool(accounts[1], { from: accounts[2] }), "ERR_ACCESS_DENIED");
+  });
+
+  it("should revert if pool address is set as zero address", async () => {
+    await expectRevert(oracle.setLiquidityPool(ZERO_ADDRESS), "ERR_ZERO_POOL_ADDRESS");
+  });
+
+  it("should set pool address", async () => {
+    const receipt = await oracle.setLiquidityPool(accounts[1]);
+    expect(await oracle.liquidityPool.call()).to.be.equal(accounts[1]);
+    expectEvent(receipt, "PoolAddressUpdate", {
+      _liquidityPool: accounts[1]
+    });
+  });
+
+  it("should revert if observations are not written by pool", async () => {
+    await expectRevert(oracle.write(price0, price1), "ERR_INVALID_POOL_ADDRESS");
+  });
+
+  it("should set initial observations", async () => {
+    const receipt = await oracle.write(price0, price1, { from: pool });
+
+    const timestamp = await latest();
+    const lastCumulativePrice0 = price0.mul(timestamp);
+    const lastCumulativePrice1 = price1.mul(timestamp);
+
+    expectEvent(receipt, "ObservationsUpdated", {
+      ema0: price0,
+      ema1: price1,
+      blockNumber: await latestBlock(),
+      timestamp,
+      lastCumulativePrice0: lastCumulativePrice0.toString(),
+      lastCumulativePrice1: lastCumulativePrice1.toString()
+    });
+  });
+
+  it("should update observations", async () => {
```

Comment:
how is an update with the same price info a good test?
we should check that the pricing is updated accordingly if the prices change.

Reply:
Added the tests for all the price values from this sheet https://docs.google.com/spreadsheets/d/1iQ9u0AxQO5tNnXBcyWoa2YZvEPpmNssn8918Je7uiEY/edit#gid=0
and checked if the EMA is as expected.